[1단계 - 페이먼츠 모듈] 해삐(박소희) 미션 제출합니다.

thgml05 opened the pull request:
안녕하세요, 시지프!😀 해삐입니다.
이번 미션은 주렁과 함께 페어 프로그래밍으로 진행하였습니다!!
아직 배포나 모듈화에는 익숙하지 않아서 미숙한 부분도 많지만,
리뷰를 통해 많이 배워가겠습니다! 잘 부탁드립니다 😊


## 🎯 페이먼츠 모듈

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계 - Module

- 재사용 가능한 모듈화된 컴포넌트를 개발하고 npm에 배포할 수 있다.
- 재사용 가능한 커스텀 훅을 개발하고 npm에 배포할 수 있다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 문서화하고 테스트할 수 있다.
- React Testing Library를 사용하여 컴포넌트와 커스텀 훅에 대해 폭넓은 테스트를 작성할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항(모달 컴포넌트 및 커스텀 훅 개발, npm 배포)을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] Storybook과 RTL 테스트 케이스를 모두 작성했나요?
- [x] 코드 컨벤션을 지키고, 에러 핸들링을 고려했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
- [x] 배포한 npm 패키지가 정상적으로 설치 및 작동하나요?
  - 배포 npm 링크: [모달](https://www.npmjs.com/package/jurunghappy-hooks)/[훅](https://www.npmjs.com/package/jurunghappy-hooks) 
- [x] Storybook 데모 페이지에 정상적으로 접근할 수 있나요?
  - Storybook 링크: [스토리북](https://68131f436576e04c1cfdd91c-jzldlrrpqe.chromatic.com/)

### 리뷰 요청 & 논의하고 싶은 내용

### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
**1. 모달 모듈**

모듈을 만들 때 사용자에게 어느 정도의 유연성을 제공할지에 대해 페어와 많이 고민했던 것 같습니다. 기능적으로나 구조적으로 다양한 기준이 존재할 수 있지만, 저희는 이번 미션에 초점을 맞춰서 모달 모듈을 구현해보았습니다. 모달은 배경, 컨테이너, 타이틀, 닫기 버튼의 네 가지 요소가 항상 동일하게 들어가야 했기 때문에, 이들을 기본 레이아웃으로 고정해 제공하고자 했습니다. 그리고 그 외의 내용 아래 부분에 추가될 내용은 사용자마다 달라질 수 있다고 생각해서 해당 영역은 children prop을 통해 유연하게 확장할 수 있도록 설계하였습니다. 이전 미션들에서도 모달을 자주 사용했지만, 이번 미션을 통해 단순히 기능으로서 사용하는 것만이 아니라 사용자에게 어떤 형태로 모듈을 제공하고 싶은지에 대해 생각해볼 수 있었습니다. 모달의 레이아웃을 어느 정도까지 고정하는지에 따라 활용도가 크게 달라질 수 있을 것 같습니다.

**2. 훅 모듈**

훅을 만들 때에도 역시나 어디까지 분리할지 고민이 되었습니다. 공통 로직을 최대한 분리할지, 각 카드 입력값 별로 만들지, 상태는 어떤 걸 제공할지에 대해 이야기를 나누었고, 저희는 각 카드 입력값 별로 훅을 만들어보았습니다. 카드를 발급할 때 들어가는 입력값이 왠만하면 카드 번호, 브랜드, 유효 날짜, CVC, 비밀번호가 고정되어 있을거라고 생각했기 때문에 브랜드를 제외한 각각의 입력값에 따른 훅을 개별로 만들어서 제공해도 될 것 같다고 판단했습니다. 각 입력 값에 따라 필요한 훅만 import해서 사용하면 되기 때문에 직관적이고 편리할 것 같다고 생각했습니다. 이전 페이먼츠 미션에서는 훅에 불필요한 상태값을 많이 선언하였는데 이번 모듈화 미션을 진행하면서 재사용성과 필요한 상태값만을 가진 훅을 다시 한번 만들어 볼 수 있었습니다.

**3. 배포 & 설치**
이번 미션을 통해 모듈 설계부터 npm 배포까지 처음으로 경험해보았습니다. 배포 과정에서 예상치 못한 오류들을 여러 번 마주하면서 npm 배포가 생각보다 어렵게 느껴졌던 것 같습니다..!🥲 특히 package.json을 비롯한 설정 파일들의 역할과 의미를 좀 더 이해해보는 시간이 필요할 것 같습니다. 이번 미션를 통해 어떤 절차와 명령어들을 통해 배포가 진행되는지 전체적으로 배워갈 수 있었습니다. 

**4. emotion CSS**
module css를 처음에 사용하였지만 배포 오류 문제를 해결하지 못해서 emotion css를 처음 사용해보았습니다. 태그 이름을 커스텀 할 수 있는 점과 태그 내부에 className이 길게 들어가지 않아서 가독성 측면에서도 깔끔하게 느껴진다는 점이 좋았습니다. 스타일링에 조건을 추가하는 코드는 조금 복잡해보일 수 있다고 생각하지만 실제 컴포넌트에서 사용할 때는 깖끔한 가독성을 유지할 수 있을 것 같습니다. 


### 2) 이번 리뷰를 통해 논의하고 싶은 부분
**1. 훅 모듈 유연성 정도**
저희 페어는 각 입력값에 따라 훅을 개별로 생성하여 배포하였습니다. 각각의 훅으로 관리하는 게 직관적이고 편리하다고 생각해서 지금과 같이 설계해보았는데 usePassword와 useCVC의 경우 로직이 유효성 기준을 제외하고는 동일하기 때문에 같은 훅으로 통합시키는 방법도 가능하다고 생각합니다. 하지만 유지보수를 생각한다면 개별 훅으로 두는 것이 더 관리하기 편하지 않을까 하는 생각도 듭니다. 공통되는 로직을 가진 훅은 하나로 통합시키는 게 더 나은지 아니면 개별이 더 좋을지에 대한 시지프의 의견이 궁금합니다! 추가로 훅에서 개선될 부분에 대한 피드백 받고 싶습니다!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 모달 컴포넌트의 유연성

- 다양한 위치(중앙, 하단 등)와 내용을 prop으로 받아 모달을 유연하게 구성할 수 있는가?
- 열기, 닫기, 확인 등의 사용자 정의 이벤트 핸들러가 자연스럽게 동작하는가?
- Storybook을 통해 다양한 시나리오를 명확히 테스트했는가?

### 2. 커스텀 훅의 역할과 모듈화

- 카드 정보 입력 유효성 검사 훅이 각각 명확한 책임을 가지도록 분리되어 있는가? 
- 훅이 유효성 검사 결과와 에러 메시지를 명확히 반환하는가?
- 독립적인 테스트 케이스를 작성하여 정상/비정상 입력 모두를 폭넓게 검증했는가?

### 3. 모듈 배포
- npm 배포를 위한 설정이 적절히 구성되어 있는가?

wmakerjun committed:
docs: PR 템플릿 개선

AHHYUNJU committed:
docs: README 작성

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
docs: 컴포넌트와 훅 리드미 분리
Co-authored-by: thgml05 <aisohii@naver.com>

AHHYUNJU committed:
feat: title을 prop로 받는 모달 컴포넌트 생성
Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: 모달 열기 버튼 구현

AHHYUNJU committed:
feat: 모달 닫기 버튼 구현

AHHYUNJU committed:
feat: 모달 header css 구현

AHHYUNJU committed:
refactor: 모달의 유연성을 위한 구조 개선

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
docs: README 최신화

AHHYUNJU committed:
fix: 모달 position에 따른 css 오류 수정

AHHYUNJU committed:
feat: 모달 컴포넌트 배포 및 설치

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: 모달 스토리북 구현

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: useCvcNumber 훅 구현

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: usePassword 훅 구현

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: useExpiryDate 훅 구현

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
docs: README 수정

AHHYUNJU committed:
feat: useCardNumbers 훅 구현

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: 유효성 검사 커스텀 훅 배포

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
test: 유효성 검사를 위한 커스텀 훅 RTL 테스트 코드 추가

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
refactor: module.css를 emotion으로 변경

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
fix: 불필요한 파일, 주석, 코드 제거

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
docs: README update

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
fix: 유효성 검사를 위한 커스텀 훅에서 빈 입력 시 에러 메시지를 지우도록 수정

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: 유효성 검사를 위한 커스텀 훅 재배포

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

AHHYUNJU committed:
feat: 유효성 검사를 위한 커스텀 훅의 RTL 테스트 성공 케이스 구현

AHHYUNJU committed:
feat: App.tsx에 cvc 유효성 커스텀 훅 추가

Co-authored-by: thgml05 <thgml05@users.noreply.github.com>

thgml05 committed:
docs : 모듈 설명 리드미 추가 및 기능 목록 리드미 수정

thgml05 committed:
chore : 모듈 재설치

thgml05 committed:
fix : 유효 날짜 유효성 검사 오류 수정

thgml05 committed:
chore : 모듈 재설치

thgml05 committed:
fix : 모듈 css 오류 수정

thgml05 committed:
refactor : 불필요한 주석 제거

woowa-euijinkk commented on components/src/vite-env.d.ts:
```diff
@@ -1 +1,5 @@
 /// <reference types="vite/client" />
+declare module "*.module.css" {
+  const classes: { [key: string]: string };
+  export default classes;
+}
```
이 코드는 어떤건가요?~

thgml05 replied:
타입스크립트가 ```<div className={styles.modalContainer} />``` 와 같은 styles를 사용한 module.css의 코드를 이해하도록 하기 위한 설정이라고 알고 있습니다. 이번 미션에서는 emotion.css를 사용하였기 때문에 불필요한 파일이라 삭제하였습니다.

woowa-euijinkk commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,37 @@
+} from './Modal.styles.ts';
+interface ModalPropsType {
+  isModalOpen: boolean;
+  position: string;
```
타입을 더 좁혀야 하지 않을까요?~

thgml05 replied:
'center'와 'bottom' 외의 값은 현재 컴포넌트에서 처리할 수 없기 때문에, 타입을 'center' | 'bottom'으로 제한하는 것이 더 안전한 것 같습니다. 'center'와 'bottom'으로 타입 제한하였습니다.
커밋 : [4bfcb15](https://github.com/woowacourse/react-modules/pull/116/commits/4bfcb1583c03bca601459072f6c6e94a87efa17a)

woowa-euijinkk commented on components/src/lib/Modal.tsx:
```diff
@@ -0,0 +1,37 @@
+  CloseButton,
+} from './Modal.styles.ts';
+interface ModalPropsType {
+  isModalOpen: boolean;
```
이미 컴포넌트 이름이 Modal 이기 때문에, 이름이 `isModalOpen` 으로 지을 필요는 없을 것 같아요!

woowa-euijinkk replied:
이런 네이밍을 잘 하기 위해서 다른 디자인 시스템 라이브러리의 이름을 많이 참고해보아요.

thgml05 replied:
디자인 시스템 라이브러리에서 어떻게 네이밍을 하고 있는지 살펴보았습니다. isOpen, open, opened 등 여러 네이밍이 있었지만, 그 중에서도 boolean 값을 명확하게 나타내는 isOpen으로 네이밍 수정하였습니다!

woowa-euijinkk commented on components/src/lib/Modal.styles.ts:
```diff
@@ -0,0 +1,60 @@
+  display: flex;
+  justify-content: center;
+  z-index: 10;
+  visibility: hidden;
+  visibility: ${({ isModalOpen }) => isModalOpen === true && 'visible'};
```
합쳐서 쓰면 더 명확하지 않을까 싶어요

thgml05 replied:
수정하였습니다!
커밋 : [9e9db9a](https://github.com/woowacourse/react-modules/pull/116/commits/9e9db9ae708be6344ee34b48de3b94e6824fc935)

woowa-euijinkk commented on components/src/lib/Modal.styles.ts:
```diff
@@ -0,0 +1,60 @@
+  color: #000000;
+  z-index: 2;
+  position: relative;
+  width: 304px;
+  height: 216px;
```
명확한 px 단위로 표현되어서, 유연하지 않은 것 같습니다. 더 잘하려면 어떻게 해야할까요?

thgml05 replied:
창에 크기에 맞춰서 크기가 유동적으로 변하도록 % 단위를 사용해보았습니다. min-width와 min-height 속성도 추가하여 창이 너무 작아질 때는 최소한의 크기를 유지하도록 설정하였습니다. 
커밋 : [1eb1f9b](https://github.com/woowacourse/react-modules/pull/116/commits/1eb1f9b838d79de9e8f060292eb4637d8f810322)

woowa-euijinkk commented on components/src/lib/Modal.styles.ts:
```diff
@@ -0,0 +1,60 @@
+  position: relative;
+  width: 304px;
+  height: 216px;
+  width:  ${({ position }) => position === 'bottom' && '100%;'}
+  border-radius:  ${({ position }) => position === 'bottom' && '10px 10px 0 0;'}
```
position 에 따른 스타일을 더 깔끔하게 표현할 수 없을까요?
코드 중복이 많고, position 별로 어떤 스타일을 가질지 더 명확하게 알 수 있으면 좋겠어요

thgml05 replied:
각 position 조건 별로 적용되어야 하는 스타일을 한번에 확인할 수 있도록 수정해보았습니다!
커밋 : [c758bfd](https://github.com/woowacourse/react-modules/pull/116/commits/c758bfdafe562c6478b42b0b09c7531627a70953)

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+  errorMessage: string;
+};
+
+const initialDate = {
```
왜 이름이 initialDate 에용?~

thgml05 replied:
error 객체인데 initialDate로 이름을 지었네요..! 네이밍을 잘못 작성한 것 같습니다! initialError로 수정하였습니다!!

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+  errorMessage: "",
+};
+
+export default function useCardNumbers(): ValitationResult {
```
훅 사용이 불편해보여요. 해삐 코드의 react-payments 미션에 한번 적용해보고 어떤 것이 문제인지 생각해보시겠어요?~

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+  errorMessage: "",
+};
+
+export default function useCardNumbers(): ValitationResult {
```
훅 사용이 불편해보이는데요. 해삐의 react-payments 에 적용해보고, 어떤 점이 이상한지 한번 생각해보시겠어요?~

thgml05 replied:
아직은 불편한점을 발견하지 못했는데 이 부분은 조금 더 생각해보도록 하겠습니다!!😀

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+    });
+  };
+
+  const validate = (value: string, index: number) => {
```
validate 함수가 update 를 수행하는게 이상합니다!
이 부분 전체적으로 수정이 필요하겠어요

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+    Array.from({ length: 4 }, () => initialDate)
+  );
+
+  const updateCardNumber = (
+    index: number,
+    isError: boolean,
+    message: string
+  ) => {
+    setError((prev) => {
+      prev[index] = {
+        isValidate: isError,
+        errorMessage: message,
+      };
+      return prev;
+    });
+  };
```
updateCardNumber 인데 setError 만 수행하는게 이상합니다. 역할이 이상해보여요!

thgml05 replied:
이 부분도 네이밍이 잘못된 것 같습니다. setError 작업이 반복적으로 사용되기도 하고 복잡해보여서 함수로 분리하였습니다! 이름은 updateError로 수정하였습니다!

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+  );
+
+  const updateCardNumber = (
+    index: number,
+    isError: boolean,
+    message: string
```
인자가 3개 이상이라 순서 기억이 어렵습니다. 어떻게 해결하면 좋을까요?

woowa-euijinkk replied:
또한 isError 가 false 여도, message 를 넣어줘야 하는게 이상합니다. 어떻게 해소할 수 있을까요?
discriminated union 을 학습해보아요.

thgml05 replied:
인자가 3개인 코드에 대해서는 넘겨 받는 매개변수를 객체로 수정해보았습니다.
순서대로 어떤 값이 넘겨지는지는 key 값으로 명확하게 보이게 되었다고는 생각하지만 시지프가 말씀해주신 "순서 기억이 어렵다"는 말에 부합한 코드인지는 정확히 모르겠습니다..! "넘겨주는 순서를 명확히 보여준다"는 의미로 해석해서 코드를 수정해보았는데 혹시 맞을까요..? 아니면 조금 더 순서를 미리 알 수 있는 해결법을 찾아야 하는걸까요? 🧐

discriminated union이라는 개념을 처음 알게 되었습니다. 
해당 개념을 활용해서 아래와 같이 타입 정의해보았습니다.
```js
type UpdateErrorArgs =
  | { index: number; isValid: true; errorMessage: string }
  | { index: number; isValid: false };
```
이 타입의 isValid를 활용해서 isValid가 false일 때는 ''를 errorMessage로 세팅할 수 있도록 코드 수정했습니다.
```js
  const updateError = (args: UpdateErrorArgs) => {
    setError((prev) => {
      const updated = [...prev];
      updated[args.index] = {
        isValid: args.isValid,
        errorMessage: args.isValid ? args.errorMessage : '',
      };
      return updated;
    });
  };
```
이 방법을 활용하니까 false일 때 굳이 메시지를 넘겨주지 않아도 기본 세팅이 되어서 좋은 것 같습니다!

커밋 : [ac2d45e](https://github.com/woowacourse/react-modules/pull/116/commits/ac2d45e47f59ad9296f2e4625286a611569e0a49)

woowa-euijinkk replied:
discriminated union 잘 적용해주셨네요! 👍 좋습니다~

woowa-euijinkk replied:
> 순서대로 어떤 값이 넘겨지는지는 key 값으로 명확하게 보이게 되었다고는 생각하지만 시지프가 말씀해주신 "순서 기억이 어렵다"는 말에 부합한 코드인지는 정확히 모르겠습니다..!

네 부득이 하게 인자가 3개 이상이어야 하는 경우, 객체로 쓰면 좋다는 의미였습니다. 순서를 기억하지 못하니 말이죠.

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.ts:
```diff
@@ -0,0 +1,71 @@
+      return newNumbers;
+    });
+
+    if (value === "") {
+      updateCardNumber(index, false, "");
+      return;
+    }
+
+    if (!/^\d*$/.test(value)) {
+      updateCardNumber(index, true, CARD_NUMBER_RULE.NOT_A_NUMBER);
+      return;
+    }
+    if (value.length < CARD_NUMBER_RULE.MAX_LENGTH) {
+      updateCardNumber(index, true, CARD_NUMBER_RULE.INVALID_LENGTH_ERROR);
+      return;
+    }
```
각 조건문은 함수로 분리되면 더 명확할 것 같습니다.

thgml05 replied:
validate.ts 파일에 함수로 분리하였습니다! 재사용성도 고려되었고 조건문의 역할도 명확하게 보이는 것 같습니다.
커밋 : [8897148](https://github.com/woowacourse/react-modules/pull/116/commits/88971481d488f602fe3941ce1d06d218da16bfad)

woowa-euijinkk replied:
잘 분리해주셨네요~

woowa-euijinkk commented on hooks/src/lib/useCardNumbers.test.ts:
```diff
@@ -0,0 +1,36 @@
+  const { result } = renderHook(() => useCardNumbers());
+
+  act(() => {
+    result.current.validate("1234", 0);
```
만약 이걸 TDD 로 코드를 짜셨다면, 절대 이름을 validate 로 짓지 않았으리라 생각합니다. 
TDD 의 장점! -> "설계와 인터페이스에 집중해서 코드를 짤 수 있도록 해준다."

너무 구현에 집중하여 코드를 짜시지 않았나 생각이 듭니다. 그런 방식에서 벗어나려면, interface 를 먼저 짜고 코드를 짜면 도움이 될거에요. TDD 를 하셔도 되고요 ㅎㅎ

thgml05 replied:
네!ㅎㅎ 저도 매 미션을 진행하면서 구현에만 좀 초점을 맞춰서 코드를 작성하고 있는 것 같다는 느낌을 받았습니다. 말씀해주신 것처럼 인터페이스와 설계에 먼저 집중하는 연습이 확실히 필요하다고 느꼈습니다! 시지프가 말씀해주신 interface와 TDD 방식도 참고해보겠습니다. 감사합니다!!

woowa-euijinkk commented on hooks/src/lib/useExpiryDate.ts:
```diff
@@ -0,0 +1,111 @@
+    });
+  };
+
+  const validate = (value: string, dateName: string) => {
```
dateName 이름이 이상하네요 ㅎㅎ + string 말고 더 타입을 좁혀보아요.

thgml05 replied:
dateName(날짜의 이름) 변수명의 의미가 모호했던 것 같습니다. ‘month’와 ‘year’를 담고 있다는 의미가 조금 더 드러날 수 있는 dateUnit으로 수정하였습니다. 타입도 ‘month’와 ‘year’로 타입을 좁혔습니다. 
커밋 : [436be70](https://github.com/woowacourse/react-modules/pull/116/commits/436be704561ec33e779242348eef5dfe3672d484)

woowa-euijinkk commented on hooks/src/lib/useExpiryDate.ts:
```diff
@@ -0,0 +1,111 @@
+        return;
+      }
+      if (value.length < EXPIRYDATE_RULE.DATE_MAX_LENGTH) {
+        updateDate(0, true, EXPIRYDATE_RULE.INVALID_MONTH_LENGTH_ERROR);
+        return;
+      }
+      if (
+        Number(value) < EXPIRYDATE_RULE.DATE_MONTH_MIN ||
+        Number(value) > EXPIRYDATE_RULE.DATE_MONTH_MAX
+      ) {
+        updateDate(0, true, EXPIRYDATE_RULE.INVALID_MONTH);
+        return;
+      }
+      updateDate(0, false, '');
+    }
+
+    if (dateName === 'year') {
+      if (value.length > EXPIRYDATE_RULE.DATE_MAX_LENGTH) return;
+      setDate((prev) => ({ ...prev, year: value }));
+
+      if (value === '') {
+        updateDate(1, false, '');
+        return;
+      }
+
+      if (!/^\d*$/.test(value)) {
+        updateDate(1, true, EXPIRYDATE_RULE.YEAR_IS_NOT_A_NUMBER);
+        return;
+      }
+      if (value.length < EXPIRYDATE_RULE.DATE_MAX_LENGTH) {
+        updateDate(1, true, EXPIRYDATE_RULE.INVALID_YEAR_LENGTH_ERROR);
+        return;
+      }
+      if (Number(value) < EXPIRYDATE_RULE.DATE_YEAR_MIN) {
+        updateDate(1, true, EXPIRYDATE_RULE.INVALID_YEAR);
+        return;
+      }
+      updateDate(1, false, '');
+    }
+  };
```
이렇게 분기로 짜지 말고, 그냥 함수 자체를 나누면 안되나요?~
validateMonth, validateYear

thgml05 replied:
validateMonth와 validateYear 함수로 분리하였습니다. 분기로 나누는 것 보다 훨씬 직관적으로 역할이 보이는 것 같습니다!_

woowa-euijinkk commented on hooks/src/lib/useCvcNumber.test.ts:
```diff
@@ -0,0 +1,34 @@
+    result.current.validate("12");
+  });
+
+  expect(result.current.error.errorMessage).toBe(
+    "CVC는 3자리로 입력해 주세요."
+  );
```
깨지기 쉬운 테스트네요. 에러메시지가 바뀌면, 이 테스트가 바로 깨질텐데 그게 맞을까요?~

woowa-euijinkk reviewed (changes requested):
해삐 1단계 구현하시느라 고생 많으셨어요.
npm 배포가 처음이라 어려우셨을텐데, 잘 극복하셨군요! 👍 
전반적으로 잘해주셨는데요. 몇가지 코멘트 전달드리겠습니다!

## 1. 함수의 인터페이스 고민하기

함수명, 변수명이 아쉬운게 있어요. 이것은 구현에 집중하기 때문에 나타난 현상이라고 생각하는데요. interface 를 먼저 구상하고 코드를 짜신다면,  더 나은 이름을 지으실 수 있을거에요.

가령, 이 hook 을 구현하기 전에 쓰는 것을 상상하는 거에요. 그리고 아래처럼 interface 만 먼저 짠다면, validate 라는 이름이 나올 순 없겠다 싶어요. 
`const { cardNumbers, handleChange } = useCardNumbers()`

## 2. storybook 개선
storybook 에서 모달을 띄우고 닫는 것 까지 볼 수 있어야할 것 같아요.
그래야, 어떤 애니메이션이 있는지, 닫히는 동작이 잘 되는지 등을 확인할 수 있을 것 같아요.

## 3. UX 개선
backdrop 을 클릭했을 때 닫히는 코드가 없군요. 이런 부분도 추가되면 라이브러리로서 더 좋겠네요.

## 4. hook 의 공통 로직 분리
질문 주신 것에 답변 드리겠습니다!
layer 를 두면 어떨까 싶습니다. 공통 로직을 분리한 훅이 있고, 그것을 개별 훅에서(usePassword, useCVC) 사용하여, export 하면 되지 않을까 싶습니다.

개별 훅, 공통 훅, 이런 부분은 앞으로도 계속 고민해갈 텐데요. 해삐만의 기준을 세워보시면 좋겠네요! 많이 고민해보아요.
상황에 따라 다를텐데요.
이번 상황은, useCardNumbers 나 useExpiryDate 는 별개 로직이 많이 들어가기 때문에 분리가 되어야겠죠. 그런데, usePassword나 useCVC는 또 개별 훅으로 나오지 않는다면, 사용 시 매우 헷갈릴거같아요. 왜 이건 개별 훅이고, 이건 공통훅인가.. 이처럼 "구현" 입장에서 생각하는게 아니라, "사용자 인터페이스" 입장에서 생각해보시면 좋아요.

imakerjun reviewed (commented):
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

wmakerjun commented:
@coderabbitai Based on commit history, identify the **recurring work patterns** this developer uses to solve problems.

Focus on these two key questions:

1. **What consistent workflow or cycle does the developer follow?**
    
    – For example: feature → styling → testing. Explain with specific commit examples.
    
2. **What does this pattern suggest about the developer’s mindset or strategy?**
    
    – e.g., feature-first, cleanup-later, test-driven, etc.
    

Structure the result with a **clear title + example + interpretation**, and provide the answer **in Korean**.

thgml05 committed:
refactor: 모달 오픈 상태 변수명 수정

isModalOpen 상태 변수명을 isOpen으로 변경

thgml05 committed:
refactor: position 타입 수정

string 타입을 'center' | 'bottom' 타입으로 수정

thgml05 committed:
refactor: ModalBackground 스타일 visibility 조건문 수정

thgml05 committed:
refactor: color와 z-index 스타일 속성 값 변수화

thgml05 committed:
refactor: ModalContainer의 position 조건별 스타일 구조 개선

'center'와 'bottom'에 대한 스타일을 깔끔한 구조로 개선

thgml05 committed:
refactor: 모달 너비와 높이 단위 조정

- 유동적으로 변할 필요성이 있는 너비와 높이는 % 단위로 변경
- 필요한 경우 최소 너비와 최소 높이 추가

thgml05 committed:
refactor: 역할에 맞지 않는 변수와 함수명 수정

- initialDate을 initialError로 수정
- updateCardNumber와 updateExpiryDate를 updateError로 수정

thgml05 committed:
refactor: useExpiryDate 훅 수정

- dateName 이름 dateUnit으로 변경
- dateUnit 타입 'month' | 'year'로 수정
- 월/년도 validate 함수 분리

thgml05 committed:
refactor: 각 훅의 validate 함수에서 update 로직 별개의 함수로 분리

thgml05 committed:
refactor: 에러 메시지와 유효성 검증 규칙 상수 파일로 분리

- errorMessage.ts : 예외 처리 메시지
- validationRule.ts : 유효성 검증 규칙

thgml05 committed:
refactor: 각 훅에서 updateError에 매개변수 전달 시 errorMessage 전달 여부 수정

thgml05 committed:
refactor: 각 훅 validate 함수의 조건문 함수로 분리

thgml05 committed:
refactor: 공통 ErrorType과 initialError 객체 분리

thgml05 committed:
chore: css module 설정 파일 삭제

thgml05 committed:
refactor: 각 훅의 update 함수handle로 시작하도록 함수명 수정

thgml05 committed:
feat: backdrop 클릭 시 모달창 닫히는 props 추가

thgml05 committed:
feat: Modal 스토리북 'center'와 'bottom'일 때 수동 동작 테스트 코드 추가

thgml05 committed:
docs: modal과 훅 리드미 설명 수정

thgml05 committed:
chore: modal과 hooks 모듈 재배포 및 설치

thgml05 commented on components/src/lib/Modal.styles.ts:
```diff
@@ -0,0 +1,60 @@
+  padding: 20px;
+  border-radius: 16px;
+  color: #000000;
+  z-index: 2;
```
z-index이 이루어지고 있는 주체가 명확히 보이도록 값을 변수화 하였습니다! 
커밋 : [6143f62](https://github.com/woowacourse/react-modules/pull/116/commits/6143f6218bd84452ae8eaaca287932d234846b3d)

woowa-euijinkk reviewed (approved):
코멘트 많이 드렸는데, 수정하시느라 고생 많으셨어요. 많이 배우는 계기가 되었으면 좋겠네요!

제가 수정 방향을 알려드리지 않고, 질문을 드린 부분도 많았는데요. 해삐가 의도를 잘 이해하고, 알맞은 방향으로 잘 수정해주셔서, 추가로 코멘트 드릴 부분이 없네요. 
잘하셨어요!

다만, 이번 미션은 기본적인 구현에 충실해주셨는데요. 다음 미션에서, 한단계 더 나아가서, 심화된 구현도 신경 써보시면 좋을 것 같네요! 잘하고 계시니 여러 도전들 해보아요!

고생하셨습니다! step2 에서 봬요.

wmakerjun commented:
@coderabbitai
You are a senior React engineer.

# Task
From this pull request’s code, identify **exactly one recurring code smell or anti-pattern** whose removal will most improve long-term maintainability and correctness.

# How to respond
- **Pattern name**  
- **Where it appears** (files + line ranges or clear examples)  
- **Why it hurts** (short reason + link to React/JS best-practice doc if relevant)  
- **How to fix** (concise code snippet or pseudo-diff)  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style
Respond in bullet points, no extra tips or unrelated suggestions.

# Language
Please write your response **in Korean**.

thgml05 commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one place** where the code is written in an imperative style, and explain how to make it **more declarative**.

# How to respond  
Start your response with the following markdown title:  
`# 1. ✨ 선언형 UI로서 개선할 피드백 포인트`

Then respond with:  
- **Imperative pattern name or symptom**
- **Where it appears** (files + line ranges or clear examples)  
- **Why it’s not declarative**  
- **AS-IS**: _Show the current (imperative) code snippet_  
- **TO-BE**: _Show the rewritten (declarative) version_  
- **Why this change improves React code quality** (≤ 2 sentences)

# Style  
Respond in bullet points, and include only one clear improvement point.

# Language  
Please write your response **in Korean**.

thgml05 commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one state value** that can be **derived from other state or props**, and explain how removing it would simplify the code.

# How to respond  
Start your response with the following markdown title:  
`# 2. 📉 상태 최소화를 위한 개선 포인트`

Then respond with:  
- **Derived state name**  
- **Where it appears** (files + line ranges or variable name)  
- **Why it’s redundant**  
- **AS-IS**: _Show the redundant state and its logic_  
- **TO-BE**: _Show how to derive it from existing values_  
- **Why this is the highest-leverage fix** (≤ 2 sentences)

# Style  
Respond in bullet points, and focus on one state-related improvement.

# Language  
Please write your response **in Korean**.

thgml05 commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

thgml05 commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one recurring mistake, anti-pattern, or code smell** that appears in multiple places and seems to reflect the author's coding habit.  
Explain what it is, why it may keep happening, and how the author can recognize and change this pattern.

# How to respond  
Start your response with the following markdown title:  
`# 4. 🔁 반복적으로 나타나는 안티패턴 또는 코드 습관 피드백`

Then respond with:
- **Pattern or smell name**  
- **Where it appears repeatedly** (files + line ranges or variable/component patterns)  
- **Why it likely keeps happening** (assumed thought process or misconception)  
- **AS-IS**: _Example of current usage or structure_  
- **TO-BE**: _Refactored or improved structure with reasoning_  
- **How the author can self-check next time** (one reflection or checklist question)

# Style  
Respond in bullet points, focused on identifying **a single recurring behavioral pattern**.

# Language  
Please write your response **in Korean**.

thgml05 commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, identify **exactly one component** whose responsibility is unclear or mixed, and explain how to refactor it to **separate concerns more clearly**.

# How to respond  
Start your response with the following markdown title:  
`# 3. 🧩 컴포넌트 책임 분리를 위한 리팩토링 포인트`

Then respond with:  
- **Component name or file**  
- **Where the responsibility is unclear**  
- **Why this hurts reusability and maintainability**  
- **AS-IS**: _Show the current (overloaded) component logic_  
- **TO-BE**: _Show a refactored structure or extraction example_  
- **Why this is the most impactful structural change** (≤ 2 sentences)

# Style  
Respond in bullet points, and provide clear before/after contrast.

# Language  
Please write your response **in Korean**.

thgml05 commented:
@coderabbitai  
You are a senior React engineer.

# Task  
From this pull request’s code, review whether the `validate` function—currently defined inline or within the component—should be extracted into a utility function.  
If it is already well-placed and scoped, affirm that choice.  
If extracting it would improve readability, testability, or reuse without adding overhead, suggest one possible direction.

# How to respond  
Start your response with the following markdown title:  
# ✨ 한글 제목: validate 함수 분리 여부 검토

Then respond with:
- ✅ **현재 코드가 원칙에 잘 맞는 이유** (if applicable)  
- ⚠️ **개선 여지가 있다면, 가능한 방향** (optional)  
- **AS-IS**: _Show the current code_
- **TO-BE**: _Show a refactored example_
- **스스로 점검할 수 있는 질문 하나**

# Style  
Use bullet points only. Do not force a suggestion—only provide one if it adds meaningful value.

# Language  
Please write your response **in Korean**.
